pkg/controller/direct/secretmanager/secret_controller.go

```diff
@@ -0,0 +1,381 @@
+}
+
+func UserAnnotations(secret *krm.SecretManagerSecret) map[string]string {
+	annotations := MergeMap(secret.GetAnnotations(), secret.Spec.Annotations)
+	annotations = RemoveByPrefixes(annotations, "cnrm.cloud.google.com", "alpha.cnrm.cloud.google.com")
+	return annotations
+}
+
+func UserLabels(secret *krm.SecretManagerSecret) map[string]string {
+	labels := RemoveByPrefixes(secret.GetLabels(), "cnrm.cloud.google.com")
+	labels["managed-by-cnrm"] = "true"
+	return labels
+}
+
+func (a *Adapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	u := createOp.GetUnstructured()
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("creating Secret", "name", a.id.FullyQualifiedName())
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	resource := SecretManagerSecretSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	resource.Annotations = UserAnnotations(desired)
+	resource.Labels = UserLabels(desired)
+
+	// Add metadata
+	req := &secretmanagerpb.CreateSecretRequest{
+		Parent:   a.id.Parent.String(),
+		SecretId: a.id.Secret,
+		Secret:   resource,
+	}
+	created, err := a.gcpClient.CreateSecret(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating Secret %s: %w", a.id.FullyQualifiedName(), err)
+	}
+	log.V(2).Info("successfully created Secret", "name", a.id.FullyQualifiedName())
+
+	status := &krm.SecretManagerSecretStatus{}
+	status.ObservedState = SecretManagerSecretStatusObservedState_FromProto(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.ExternalRef = a.id.AsExternalRef()
+	status.Name = created.Name
+	return setStatus(u, status)
+}
+
+func topicsEqual(desired []*krm.TopicRef, actual []*secretmanagerpb.Topic) bool {
+	externalsDesired := []string{}
+	externalsActual := []string{}
+	for _, topicRef := range desired {
+		externalsDesired = append(externalsDesired, topicRef.PubSubTopicRef.External)
+	}
+	for _, topic := range actual {
+		externalsActual = append(externalsActual, topic.GetName())
+	}
+	sort.Sort(sort.StringSlice(externalsDesired))
+	sort.Sort(sort.StringSlice(externalsActual))
+	return reflect.DeepEqual(externalsDesired, externalsActual)
+}
+
+func (a *Adapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	u := updateOp.GetUnstructured()
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("updating Secret", "name", a.id.FullyQualifiedName())
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	resource := SecretManagerSecretSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	// the GCP service use *name* to identify the resource.
+	resource.Name = a.id.FullyQualifiedName()
+
+	resource.Annotations = UserAnnotations(desired)
+	resource.Labels = UserLabels(desired)
+
+	updateMask := &fieldmaskpb.FieldMask{}
+
+	if !reflect.DeepEqual(resource.Annotations, a.actual.Annotations) {
```

Comment:
I wonder if we can infer from this which fields are immutable (or at least we don't support update), to add another "interlock" so that resources are more correct-by-design.

Reply:
Yes! And update needs more helpers in general!

Reply:
Used the protos field_behavior to find out the immutable and add a function to do proto level comparison in #2904